drivers/i2c/i2c_nrfx_twim.c

```diff
@@ -99,11 +132,32 @@ static int i2c_nrfx_twim_transfer(struct device *dev, struct i2c_msg *msgs,
 		}
 
 		res = get_dev_data(dev)->res;
+
 		if (res != NRFX_SUCCESS) {
 			LOG_ERR("Error %d occurred for message %d", res, i);
 			ret = -EIO;
 			break;
 		}
+
+		/* If concatenated messages were I2C_MSG_READ type, then
+		 * content of concatenation buffer has to be copied back into
+		 * buffers provided by user. */
+		if ((msgs[i].flags & I2C_MSG_READ)
+		    && cur_xfer.p_primary_buf == concat_buf) {
+			for (int j = i; j >= 0; j--) {
+				if (!(msgs[j].flags & I2C_MSG_READ)) {
+					break;
```

Comment:
Instead of this, you should rather check if `concat_len < msgs[j].len` (theoretically, before a set of concatenated READ messages, there could be also some with STOP flag).
Actually, when `concat_len` is used in the loop continuation condition, you don't need to check if `j >= 0`, so the loop can be:
```
while (concat_len >= msgs[j].len) {
```